### PR TITLE
Supporting Multiple Files With the Same Name

### LINKED ISSUE #315: Bug: fails to deploy two manifests with the same name
### What happened?

I have the following:

```
      ...
      ...
      - name: Deploy frontend & backend
        uses: azure/k8s-deploy@v4
        with:
          namespace: default
          manifests: |
            artifacts/frontend/deployment.yaml
            artifacts/backend/deployment.yaml
      ...
      ...
```

Since the two manifests have the same name (deployment.yaml) it deploys the same file twice. Probably, because all files are copied to the same /tmp folder, overwrite each other, and then deployed by their name only.

There's a simple workaround to use two separate steps for that, which I did and it works fine now, But I have suffered a lot before figuring out the issue.

The workaround:
```
      ...
      ...
      - name: Deploy frontend
        uses: azure/k8s-deploy@v4
        with:
          namespace: default
          manifests: |
            artifacts/frontend/deployment.yaml

      - name: Deploy backend
        uses: azure/k8s-deploy@v4
        with:
          namespace: default
          manifests: |
            artifacts/backend/deployment.yaml
      ...
      ...
```

### Version

- [X] I am using the latest version

### Runner

ubuntu-latest

### Relevant log output

```
##[debug]Evaluating condition for step: 'Deploy frontend & backend'
##[debug]Evaluating: success()
##[debug]Evaluating success:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: Deploy frontend & backend
##[debug]Loading inputs
##[debug]Evaluating: github.token
##[debug]Evaluating Index:
##[debug]..Evaluating github:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'token'
##[debug]=> '***'
##[debug]Result: '***'
##[debug]Loading env
Run azure/k8s-deploy@v[4](https://github.com/motcke/*****#step:6:4)
  
##[debug]Input manifest files: /tmp/deployment.yaml,/tmp/deployment.yaml
::group::Deploying manifests
Deploying manifests
  ##[debug]Kubectl run with command: /usr/bin/kubectl apply,-f,/tmp/deployment.yaml,/tmp/deployment.yaml,--namespace,default
  /usr/bin/kubectl apply -f /tmp/deployment.yaml,/tmp/deployment.yaml --namespace default
  service/*****-svc unchanged
  deployment.apps/*****-deploy configured
  horizontalpodautoscaler.autoscaling/*****-scaler unchanged
  secretproviderclass.secrets-store.csi.x-k8s.io/******-sync-env-vars configured
  service/******-svc unchanged
  deployment.apps/******-deploy configured
  horizontalpodautoscaler.autoscaling/******-scaler unchanged
  secretproviderclass.secrets-store.csi.x-k8s.io/******-sync-env-vars unchanged
  ##[debug]Deployed manifest files: /tmp/deployment.yaml,/tmp/deployment.yaml
  ::endgroup::
::group::Checking manifest stability
Checking manifest stability
...
...
...
```

### PR DESCRIPTION
**Pull request by jaiveerk**
Fixes #315 

Plus cleaning up how we organize files for private cluster deployment and cleaning up various file-related operations.

Added unit tests for this in manifestUpdateUtils.test.ts and modified basic integration tests to test for this scenario as well.

### CODE REVIEW

**Comment by jaiveerk** [comment]
Blocking on successful run of private cluster integration test - https://github.com/jaiveerk/k8s-deploy/actions/runs/10184782961/job/28172886784

**Comment by jaiveerk** [comment]
Waiting on successful run of private cluster integration test - https://github.com/jaiveerk/k8s-deploy/actions/runs/10184782961/job/28172886784

File: src/types/privatekubectl.ts [hunk 1]
```diff
@@ -98,63 +85,55 @@ export class PrivateKubectl extends Kubectl {
...
-export function replaceFileNamesWithBaseNames(kubectlCmd: string) {
+export function replaceFileNamesWithShallowNamesRelativeToTemp(
+   kubectlCmd: string
+) {
    let filenames = extractFileNames(kubectlCmd)
-   let basenames = filenames.map((filename) => path.basename(filename))
+   core.debug(`filenames originally provided in kubectl command: ${filenames}`)
+   let relativeShallowNames = filenames.map((filename) => {
+      const relativeName = path.relative(getTempDirectory(), filename)
+      const shallowName = relativeName.replace(/\//g, '-')
```
Comment (davidgamero) [inline src/types/privatekubectl.ts:106]: do we support windows path delimiters?

**Review by davidgamero — APPROVED** [review]
lgtm great changes

File: src/types/privatekubectl.ts (hunk 1 above)
Comment (jaiveerk) [reply src/types/privatekubectl.ts:106]: this is a great point - I'll split along the path and re-join with dashes rather than doing a string replacement!

Comment (jaiveerk) [reply src/types/privatekubectl.ts]: subbed in a split/join!

**Review by davidgamero — APPROVED** [review]

File: src/types/privatekubectl.ts [hunk 2]
```diff
@@ -43,22 +43,9 @@ export class PrivateKubectl extends Kubectl {
...
-         const filenames = extractFileNames(kubectlCmd)
-
-         const tempDirectory =
-            process.env['runner.tempDirectory'] || os.tmpdir() + '/manifests'
-         eo.cwd = tempDirectory
+         const tempDirectory = getTempDirectory()
+         eo.cwd = path.join(tempDirectory, 'manifests')
          privateClusterArgs.push(...['--file', '.'])
-
-         for (const filename of filenames) {
```
Comment (OliverMKing) [inline src/types/privatekubectl.ts:53]: It makes more sense to me to keep the moving file to the temp directory in this function

File: src/types/privatekubectl.ts [hunk 3]
```diff
@@ -98,63 +85,58 @@ export class PrivateKubectl extends Kubectl {
...
 }
 
-export function replaceFileNamesWithBaseNames(kubectlCmd: string) {
+export function replaceFileNamesWithShallowNamesRelativeToTemp(
+   kubectlCmd: string
+) {
    let filenames = extractFileNames(kubectlCmd)
-   let basenames = filenames.map((filename) => path.basename(filename))
+   core.debug(`filenames originally provided in kubectl command: ${filenames}`)
+   let relativeShallowNames = filenames.map((filename) => {
```
Comment (OliverMKing) [inline src/types/privatekubectl.ts:104]: what's the point of copying over the file in this function?

Comment (jaiveerk) [reply src/types/privatekubectl.ts:104]: they need to all be at the same level, within temp - the az cli hangs if it runs at temp root

File: src/types/privatekubectl.ts [hunk 4]
```diff
@@ -98,63 +85,58 @@ export class PrivateKubectl extends Kubectl {
...
+      const relativeName = path.relative(getTempDirectory(), filename)
+
+      const relativePathElements = relativeName.split(path.sep)
+
+      const shallowName = relativePathElements.join('-')
```
Comment (OliverMKing) [inline src/types/privatekubectl.ts:109]: why are we doing something shallow instead of just using directories?

**Review by OliverMKing — APPROVED** [review]

**Pull request merged**